CRM/Core/Payment.php

```diff
@@ -1186,9 +1186,22 @@ protected function doDirectPayment(&$params) {
    * Usage:
    * Payment processors should override this function directly instead of using doDirectPayment/doTransferCheckout which are deprecated.
    * Payment processors should set and return payment_status_id (Pending if the IPN will complete it, Completed if successful).
-   * @fixme For the contribution workflow we have a contributionID, but for the event and membership workflow the contribution has not yet been created
-   *  so we can't update params directly on the contribution.  However if you return trxn_id, fee_amount, net_amount they will be set on the contribution
-   *  by those workflows.  Ideally all workflows would create a pending contribution BEFORE calling doPayment (eg. https://github.com/civicrm/civicrm-core/pull/13763 for events)
+   * @fixme Creating a contribution record is inconsistent! We should always create a contribution BEFORE calling doPayment. However currently:
+   *  * Contribution Pages: Creates a contribution before calling doPayment.
+   *  * Contribution Pages with Membership: Does NOT create a contribution. UNLESS auto-renew=TRUE in which case both a recurring contribution
+   *      and a contribution is created before calling doPayment.
+   *  * Event Registration: Does NOT create a contribution before calling doPayment - this *looks* simple to solve: https://github.com/civicrm/civicrm-core/pull/13763
+   *
+   * If we DO have a contribution ID, then the payment processor can (and should) update parameters on the contribution record as necessary.
+   *   $params['payment_status_id'] (alias for 'contribution_status_id') should be set to the ID (via pseudoconstant getkey) of Pending/Failed/Completed depending on the outcome of the payment.
+   * @fixme Note: Anything other than Pending/Completed may not get handled correctly.
+   * If we DO NOT have a contribution ID, then the payment processor should return a $params array containing:
+   *  $params[
+   *   'payment_status_id' => {as above}
+   *   'trxn_id' => {trxn ref to link to record at payment processor}
+   *   'fee_amount' => {optional fee charged by processor}
+   *   'net_amount' => {optional net amount (ie. amount minus fee))}
```

Comment:
I don't think we should encourage returning net_amount - that can be calculated & will be elsewhere if not returned

Reply:
Ideally we should be returning nothing but payment_status_id and the payment processor should be saving the other parameters.  The workflows will only save certain parameters, net_amount is one of them.  The comment states that these should only be passed if a contribution ID is not created - and it's marked as optional.  In the long run when all flows are fixed this won't matter.

Reply:
I still don't think we should document net_amount as we don't want to be recommending something that is not preferred

Reply:
Problem we've got is that it's the only way you can actually record the amount - see eg. https://github.com/civicrm/civicrm-core/blob/master/CRM/Event/Form/Registration/Confirm.php#L1002

This comment is not meant to be "how it should be done" as that requires refactoring all these workflows so contribution is created first, at which point the payproc can take care of any/all of these params.  But this is how it "has to be done" currently if you actually want to save these params.

Reply:
if net_amount is empty it will still be calculated in the BAO no ?